File: internal/scenedetect/scenedetect.go
package scenedetect

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"time"
)

// Scene represents a detected scene boundary
type Scene struct {
	Index     int     `json:"index"`
	StartTime float64 `json:"start_time"`
	EndTime   float64 `json:"end_time"`
}

// Detector handles scene detection operations
type Detector struct {
	pythonPath        string
	scenedetectScript string
}

// NewDetector creates a new scene detector instance
func NewDetector() *Detector {
    return &Detector{
        pythonPath:        "python3",
        scenedetectScript: "/root/internal/scenedetect/sd_runner.py",
    }
}

// DetectScenes detects scenes in a video file using PySceneDetect
func (d *Detector) DetectScenes(videoPath string) ([]Scene, error) {
    // Check if Python and required dependencies are available
    if err := d.CheckDependencies(); err != nil {
        return nil, fmt.Errorf("dependencies not available: %v", err)
    }

    // Create a context with timeout for scene detection (configurable, default 300s)
    detectTimeout := 300 * time.Second
    if v := os.Getenv("SCENEDETECT_TIMEOUT_SECS"); v != "" {
        if secs, err := strconv.Atoi(v); err == nil && secs > 0 {
            detectTimeout = time.Duration(secs) * time.Second
        }
    }
    ctx, cancel := context.WithTimeout(context.Background(), detectTimeout)
    defer cancel()

    // Run PySceneDetect script
    cmd := exec.CommandContext(ctx, d.pythonPath, d.scenedetectScript, videoPath)

    out, err := cmd.CombinedOutput()
    if err != nil {
        // Try to parse JSON error from the script output
        var result struct {
            Scenes []Scene `json:"scenes"`
            Count  int     `json:"count"`
            Error  string  `json:"error,omitempty"`
        }
        if json.Unmarshal(out, &result) == nil && result.Error != "" {
            return nil, fmt.Errorf("scene detection error: %s", result.Error)
        }
        return nil, fmt.Errorf("failed to run scene detection: %v; output: %s", err, string(out))
    }

    // Parse JSON output
    var result struct {
        Scenes []Scene `json:"scenes"`
        Count  int     `json:"count"`
        Error  string  `json:"error,omitempty"`
    }

    if err := json.Unmarshal(out, &result); err != nil {
        return nil, fmt.Errorf("failed to parse scene detection output: %v", err)
    }

    if result.Error != "" {
        return nil, fmt.Errorf("scene detection error: %s", result.Error)
    }

    log.Printf("Detected %d scenes in video", result.Count)
    return result.Scenes, nil
}

// CheckDependencies checks if Python, scenedetect script, and ffmpeg are available
func (d *Detector) CheckDependencies() error {
    // Check if python is available
    cmd := exec.Command(d.pythonPath, "--version")
    if err := cmd.Run(); err != nil {
        return fmt.Errorf("python not found: %v", err)
    }

    // Check if PySceneDetect script exists
    if _, err := os.Stat(d.scenedetectScript); os.IsNotExist(err) {
        return fmt.Errorf("scenedetect script not found: %s", d.scenedetectScript)
    }

    // Check if ffmpeg is available
    cmd = exec.Command("ffmpeg", "-version")
    if err := cmd.Run(); err != nil {
        return fmt.Errorf("ffmpeg not found: %v", err)
    }

    return nil
}

// ExtractKeyframes extracts keyframes for detected scenes
func (d *Detector) ExtractKeyframes(videoPath string, outputDir string, scenes []Scene) error {
    // Create keyframes directory
    if err := os.MkdirAll(outputDir, 0755); err != nil {
        return fmt.Errorf("failed to create keyframes directory: %v", err)
    }

    // Extract keyframes using ffmpeg directly
    for i, scene := range scenes {
        // Extract a keyframe from the middle of each scene
        midTime := (scene.StartTime + scene.EndTime) / 2.0

        outputPath := filepath.Join(outputDir, fmt.Sprintf("scene_%04d_keyframe.jpg", i))

        // Create a context with timeout for keyframe extraction (configurable, default 30s)
        keyframeTimeout := 30 * time.Second
        if v := os.Getenv("KEYFRAME_TIMEOUT_SECS"); v != "" {
            if secs, err := strconv.Atoi(v); err == nil && secs > 0 {
                keyframeTimeout = time.Duration(secs) * time.Second
            }
        }
        ctx, cancel := context.WithTimeout(context.Background(), keyframeTimeout)

        cmd := exec.CommandContext(ctx, "ffmpeg",
            "-ss", fmt.Sprintf("%.2f", midTime),
            "-i", videoPath,
            "-vframes", "1",
            "-q:v", "2",
            "-y",
            outputPath,
        )

        stderr, err := cmd.CombinedOutput()
        cancel() // ensure context is canceled
        if err != nil {
            log.Printf("Warning: Failed to extract keyframe for scene %d: %v\nOutput: %s", i, err, string(stderr))
            continue
        }

        log.Printf("Extracted keyframe for scene %d to %s", i, outputPath)
    }

    return nil
}

File: internal/scenedetect/scenedetect.py
#!/usr/bin/env python3

import sys
import json
import os
from scenedetect import open_video, SceneManager
from scenedetect.detectors import ContentDetector
from scenedetect.video_splitter import split_video_ffmpeg

def detect_scenes(video_path, threshold=30.0):
    """Detect scenes in a video file using PySceneDetect"""
    try:
        # Open the video
        video = open_video(video_path)
        
        # Create a scene manager and add detectors
        scene_manager = SceneManager()
        scene_manager.add_detector(ContentDetector(threshold=threshold))
        
        # Detect scenes
        scene_manager.detect_scenes(video, show_progress=False)
        
        # Get the list of scenes
        scenes = scene_manager.get_scene_list()
        
        # Convert to our format
        scene_list = []
        for i, scene in enumerate(scenes):
            start_time = scene[0].get_seconds()
            end_time = scene[1].get_seconds()
            scene_list.append({
                'index': i,
                'start_time': start_time,
                'end_time': end_time
            })
        
        return scene_list
    except Exception as e:
        raise Exception(f"Failed to detect scenes: {str(e)}")

def extract_keyframes(video_path, scenes, output_dir):
    """Extract keyframes for detected scenes"""
    try:
        os.makedirs(output_dir, exist_ok=True)
        
        # For each scene, extract a keyframe from the middle
        for scene in scenes:
            index = scene['index']
            start_time = scene['start_time']
            end_time = scene['end_time']
            mid_time = (start_time + end_time) / 2.0
            
            output_path = os.path.join(output_dir, f"scene_{index:04d}_keyframe.jpg")
            
            # Use ffmpeg to extract the keyframe
            cmd = [
                'ffmpeg',
                '-ss', str(mid_time),
                '-i', video_path,
                '-vframes', '1',
                '-q:v', '2',
                output_path,
                '-y'  # Overwrite output files
            ]
            
            # Run ffmpeg command
            import subprocess
            subprocess.run(cmd, check=True, capture_output=True)
            
        return True
    except Exception as e:
        raise Exception(f"Failed to extract keyframes: {str(e)}")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python3 scenedetect.py <video_path> [threshold] [output_dir]")
        sys.exit(1)
    
    video_path = sys.argv[1]
    threshold = float(sys.argv[2]) if len(sys.argv) > 2 else 30.0
    output_dir = sys.argv[3] if len(sys.argv) > 3 else None
    
    try:
        # Detect scenes
        scenes = detect_scenes(video_path, threshold)
        
        # If output directory is provided, extract keyframes
        if output_dir:
            extract_keyframes(video_path, scenes, output_dir)
        
        # Output results as JSON
        result = {
            'scenes': scenes,
            'count': len(scenes)
        }
        
        print(json.dumps(result))
        
    except Exception as e:
        print(json.dumps({'error': str(e)}))
        sys.exit(1)